gprofiler/utils/node.py

```diff
@@ -0,0 +1,223 @@
+    map_file_name = f"perf-{process.pid}.map"
+    src = os.path.join(proc_root, "tmp", f"perf-{pid_inside_ns}.map")
+    dest = os.path.join("/tmp", map_file_name)
+    os.symlink(src, dest)
+
+
+@lru_cache(1024)
+def _get_pid_inside_ns(pid: int) -> int:
+    with open(f"/proc/{pid}/status") as f:
+        for line in f.readlines():
+            if line.startswith("NSpid"):
+                return line.split("\t")[-1].strip("\n")
+
+
+def _generate_perf_map_wrapper(module_path: str) -> Callable:
+    def generate_perf_map():
+        debugger_url = _get_debugger_url()
+        sock = create_connection(debugger_url)
+        _load_dso(sock, module_path)
+
+    return generate_perf_map
+
+
+def _clean_up_wrapper(module_path: str, pid: int) -> Callable:
+    def clean_up():
+        debugger_url = _get_debugger_url()
+        sock = create_connection(debugger_url)
+        _stop_dso(sock, module_path)
+        shutil.rmtree(os.path.join("/tmp", "node_module"))
+        os.remove(os.path.join("/tmp", f"perf-{pid}.map"))
+
+    return clean_up
+
+
+def generate_map_for_node_processes():
+    """Iterates over all NodeJS processes, starts debugger for it, finds debugger URL,
+    copies node-linux-perf module into process' namespace, loads module and starts it.
+    After that it links it into gProfiler namespace's /tmp. This lets perf load it"""
+    processes = get_node_processes()
+    for process in processes:
+        try:
+            musl = is_musl(process)
+            node_version = get_exe_version(process)
+            node_major_version = node_version[1:].split(".")[0]
+            dest = _copy_module_into_process_ns(process, musl, node_major_version)
+            _start_debugger(process.pid)
+            pid_inside_ns = _get_pid_inside_ns(process.pid)
+            cp = run_in_ns(
+                ["pid", "mnt", "net"], _generate_perf_map_wrapper(dest), process.pid
+            )
+            if hasattr(cp, "stdout"):
+                logger.debug(cp.stdout.decode().strip())
+            if hasattr(cp, "stderr"):
+                logger.debug(cp.stderr.decode().strip())
+            if not is_same_ns(process, NsType.mnt.name):
+                _create_symlink_for_map(process, pid_inside_ns)
+        except Exception as e:
+            logger.warn(
+                f"Could not create debug symbols for pid {process.pid}. Reason: {e}"
+            )
+            logger.debug(traceback.format_exc())
+
+
+def clean_up_node_maps():
+    """Stops generating perf maps for each NodeJS process and cleans up generated maps"""
+    processes = get_node_processes()
+    for process in processes:
+        try:
+            node_version = get_exe_version(process)
+            node_major_version = node_version[1:].split(".")[0]
+            pid_inside_ns = _get_pid_inside_ns(process.pid)
+            libc = "musl" if is_musl(process) else "glibc"
+            dest = os.path.join("/tmp", "node_module", libc, node_major_version)
```

Comment:
It's best if the final file path is versioned. This way, if in next gProfiler version we make a change to the DSO, and we want existing processes in existing deployments to get the new file, then it will be re-loaded because it's a different path. Similar to how it's done with async-profiler - we encode the async-profiler tag in the DSO path.

I suggest here that we encode e.g the 7 digits git revision from the repo of which we build the DSO. This way if we commit a fix, then the path changes and the DSO is re-loaded to existing processes